### PR TITLE
Fix: Web Request native exception on finalization

### PR DESCRIPTION
**Pull request by mikhail-dcl**
**Description**

If the web request (with an attached download handler) is GC'ed (its destructor is called) it may throw exceptions and crash the app from the native environment.

```
Received signal SIGSEGV
Obtained 21 stack frames
0x00007ff6301ccde4 (Unity) tlsf_free
0x00007ff62df8423b (Unity) DynamicHeapAllocator::Deallocate
0x00007ff62df8c526 (Unity) DynamicHeapAllocator::TryDeallocate
0x00007ff62df957dc (Unity) DualThreadAllocator<DynamicHeapAllocator>::TryDeallocate
0x00007ff62df84ab2 (Unity) MemoryManager::Deallocate
0x00007ff62df8d57a (Unity) free_alloc_internal
0x00007ff62dfb6456 (Unity) DestroySingleObject
0x00007ff62e2fd7a2 (Unity) DestroyObjectHighLevel_Internal
0x00007ff62e2fd314 (Unity) DestroyObjectHighLevel
0x00007ff62ef9dec4 (Unity) DownloadHandlerTexture::DestroyTexture
0x00007ff62e2b2e0a (Unity) BackgroundJobQueue::ExecuteMainThreadJobs
0x00007ff62e364c4f (Unity) PreloadManager::WaitForAllAsyncOperationsToComplete
0x00007ff62f5e6de2 (Unity) EditorSceneManager::RestoreSceneBackups
0x00007ff62f2e15b6 (Unity) PlayerLoopController::EnterPlayMode
0x00007ff62f2f1d76 (Unity) PlayerLoopController::SetIsPlaying
0x00007ff62f2f5470 (Unity) Application::TickTimer
0x00007ff62f76fd3a (Unity) MainMessageLoop
0x00007ff62f774c10 (Unity) WinMain
0x00007ff630b5765e (Unity) __scrt_common_main_seh
0x00007ffc2cb8257d (KERNEL32) BaseThreadInitThunk
0x00007ffc2e92aa48 (ntdll) RtlUserThreadStart
```

Here: `0x00007ff62ef9dec4 (Unity) DownloadHandlerTexture::DestroyTexture`

**How to fix it?**
Ensure disposal of the web request

**The current flow:**
- `await` `WebRequest` (without disposal) in `WebRequestController`, expose `UnityWebRequest` outside
- `await` `Data Parsing` outside of it
- `Dispose`

With 2 `awaits` there is a gap: if the flow is interrupted between the first and the second then WebRequest will be abandoned and GC'ed afterwards. 

**The new flow:**
- Process both `WebRequest` and `Data Parsing Operation` in `WebRequestController` in a single flow. `WebRequest` is wrapped in `using` to prevent a leak.
- Don't expose `UnityWebRequest` outside as it's always disposed of inside the controller

It has led to the following design changes:
- `Data Processing Operation` always should be passed to the controller, so new generic constraints were introduced
- Because of that the engineering complexity has grown
- I had to introduce the adapter to hide this complexity and minimize the changes required throughout the whole code
- The good thing is that only 1 `await` is required to process the web request and parse the data - now those operations are undetachable.

### CODE REVIEW

File: Explorer/Assets/DCL/AssetsProvision/CodeResolver/WebJsCodeProvider.cs [hunk 1]
```diff
@@ -16,9 +16,7 @@ public WebJsCodeProvider(IWebRequestController webRequestController)
 
         public async UniTask<string> GetJsCodeAsync(URLAddress url, CancellationToken cancellationToken = default)
         {
-            GenericGetRequest rqs = await webRequestController.GetAsync(new CommonArguments(url), cancellationToken);
-            string text = rqs.UnityWebRequest.downloadHandler.text;
-            rqs.UnityWebRequest.Dispose();
+            string text = await webRequestController.GetAsync(new CommonArguments(url), cancellationToken).StoreTextAsync();
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/AssetsProvision/CodeResolver/WebJsCodeProvider.cs:19]: good solution

File: Explorer/Assets/DCL/MapRenderer/MapLayers/Atlas/SatelliteAtlas/SatelliteChunkController.cs [hunk 2]
```diff
@@ -73,8 +74,9 @@ public async UniTask LoadImageAsync(Vector2Int chunkId, float chunkWorldSize, Ca
             float pixelsPerUnit = texture.width / chunkWorldSize;
 
             atlasChunk.MainSpriteRenderer.enabled = true;
-            atlasChunk.LoadingSpriteRenderer.DOColor(initialColor, 0.5f).OnComplete(()=>atlasChunk.LoadingSpriteRenderer.gameObject.SetActive(false));
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/MapRenderer/MapLayers/Atlas/SatelliteAtlas/SatelliteChunkController.cs:76]: can we extract 0.5f?

File: Explorer/Assets/DCL/WebRequests/ITypedWebRequest.cs [hunk 3]
```diff
@@ -14,7 +15,7 @@ public interface ITypedWebRequest
 
     public static class TypedWebRequestExtensions
     {
-        public static UniTask SendRequest(this ITypedWebRequest typedWebRequest, CancellationToken token) =>
+        public static UniTask SendRequest<T>(this T typedWebRequest, CancellationToken token) where T : ITypedWebRequest =>
```
Comment (NickKhalow) [inline Explorer/Assets/DCL/WebRequests/ITypedWebRequest.cs:18]: good avoiding of boxing

**Review by m3taphysics — APPROVED** [review]

**Review by NickKhalow — APPROVED** [review]

**Comment by NickKhalow** [comment]
Good solution
Nice work 💚

**Pull request merged**